### PR TITLE
Validate PCR value on server side

### LINKED ISSUE #27: Handle incorrect TPM value on server

### LINKED ISSUE #28: Compare TPM values on server side + reject non-registered
Use tpm utils validate function on server

### PR DESCRIPTION
**Pull request by T0mmas**
To check this MR if it works as it should:
- Start up your TPM Simulator
- Register to the server with some PCR value. Then your PCR value will be changed to test if it works.
- Add `from tpm2_util import tpm_pcr_extend`
- In method `quote_tpm_data` change argument in called method `tpm_connect` from `use_simulator=False` to `True`
- Then append `tpm_pcr_extend(self.tpm, 0, "Hack the Planet!")` to the end of `tpm_connect` method

Then your client app should be terminated as your PCR value has changed.

### CODE REVIEW

File: server.py [hunk 1]
```diff
@@ -78,15 +80,29 @@ def set_connection_keys(self, conn):
...
             return False
 
-    def receive_tpm_data(self, conn):
-        # TODO better storing client's tpm_data
-        self.tpm_data = conn.recv(constants.SOCK_BUFFER)
-        self.tpm_data = pickle.loads(self.tpm_data)
+    def validate_check_client_tpm(self, tpm_data: tuple) -> bool:
+        data, (ec_x, ec_y), (sig_r, sig_s) = tpm_data
+        if not ecdsa_validate(ec_x, ec_y, sig_r, sig_s, data):
+            logger.info("TPM data were validated successfully.")
```
Comment (Matej4545) [inline server.py]: If the condition is True, isn't it that it was NOT successful?

File: server.py [hunk 2]
```diff
@@ -78,15 +80,29 @@ def set_connection_keys(self, conn):
...
+        tpm_data = pickle.loads(tpm_data)
         logger.debug(f"TPM data has been received.")
+        return tpm_data
 
     def preshared_tpm_value(self, conn) -> bytes:
-        self.receive_tpm_data(conn)
-        data, _, _ = self.tpm_data
+        tpm_data = self.receive_tpm_data(conn)
+        if not self.validate_check_client_tpm(tpm_data):
+            conn.send(b"0")
```
Comment (Matej4545) [inline server.py]: Maybe we could send send something descriptive, I believe now we are encoding ASCII '0' > 0x30
Maybe then OK/NOK?

File: server.py [hunk 3]
```diff
@@ -78,15 +80,29 @@ def set_connection_keys(self, conn):
...
+            raise RuntimeError("The PCR value differs from the registered PCR value.")
+        conn.send(b"111")
```
Comment (Matej4545) [inline server.py]: Same as before.

File: server.py (hunk 1 above)
Comment (T0mmas) [reply server.py]: Yes, the logger info is not correct

File: server.py (hunk 2 above)
Comment (T0mmas) [reply server.py]: Could be. Or just send raw bytes. But why to send something descriptive when it is handled based on the T/F value?

**Review by Matej4545 — APPROVED** [review]

File: server.py (hunk 2 above)
Comment (Matej4545) [reply server.py]: For demo purposes :D no reason, but it seemed easier than parsing raw bytes.

**Pull request merged**